File: sdks/js_sdk/src/secrets/factory.ts
import { FileBackend } from './backends/file.js'
import { KeyringBackend } from './backends/keyring.js'
import type { SecretsBackend } from './backend.js'

export type BackendMode = 'auto' | 'file' | 'keyring'

function fromEnv(): BackendMode {
  const v = (process.env.SW4RM_SECRETS_BACKEND || 'auto').toLowerCase()
  if (v === 'file' || v === 'keyring') return v
  return 'auto'
}

export async function selectBackend(mode?: BackendMode): Promise<[SecretsBackend, string]> {
  const m = mode ?? fromEnv()
  if (m === 'file') return [new FileBackend(), 'file']
  if (m === 'keyring') return [new KeyringBackend(), 'keyring']
  // auto
  if (process.env.CI) return [new FileBackend(), 'file']
  try {
    // Resolve keytar quickly to decide; KeyringBackend lazily imports anyway
    await import('keytar')
    return [new KeyringBackend(), 'keyring']
  } catch {
    return [new FileBackend(), 'file']
  }
}

File: sdks/rust_sdk/src/secrets/factory.rs
use crate::secrets::backend::SecretsBackend;
use crate::secrets::backends::file::FileBackend;
use crate::secrets::errors::Result;
#[cfg(not(feature = "keyring"))]
use crate::secrets::errors::SecretError;

#[derive(Debug, Clone, Copy)]
pub enum BackendMode {
    Auto,
    File,
    Keyring,
}

impl BackendMode {
    pub fn from_env() -> Self {
        match std::env::var("SW4RM_SECRETS_BACKEND").ok().as_deref() {
            Some("file") => BackendMode::File,
            Some("keyring") => BackendMode::Keyring,
            _ => BackendMode::Auto,
        }
    }
}

pub fn select_backend(
    mode: Option<BackendMode>,
) -> Result<(Box<dyn SecretsBackend>, &'static str)> {
    let mode = mode.unwrap_or_else(BackendMode::from_env);
    match mode {
        BackendMode::File => Ok((Box::new(FileBackend::new(None)?), "file")),
        BackendMode::Keyring => {
            #[cfg(feature = "keyring")]
            {
                use crate::secrets::backends::keyring::KeyringBackend;
                Ok((Box::new(KeyringBackend::new("sw4rm")), "keyring"))
            }
            #[cfg(not(feature = "keyring"))]
            {
                Err(SecretError::Backend("keyring feature not enabled; recompile with \"keyring\" feature or use file backend".into()))
            }
        }
        BackendMode::Auto => {
            if std::env::var("CI").is_ok() {
                return Ok((Box::new(FileBackend::new(None)?), "file"));
            }
            #[cfg(feature = "keyring")]
            {
                use crate::secrets::backends::keyring::KeyringBackend;
                Ok((Box::new(KeyringBackend::new("sw4rm")), "keyring"))
            }
            #[cfg(not(feature = "keyring"))]
            {
                Ok((Box::new(FileBackend::new(None)?), "file"))
            }
        }
    }
}
